Create recipe for pynumero libraries

santiagoropb opened the pull request.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pynumero_build```) and found it was in an excellent condition.

marcelotrevisani commented on recipes/pynumero_build/meta.yaml:
```diff
@@ -0,0 +1,42 @@
+build:
+  number: 0
+  skip: true  # [win]
+  skip: true  # [osx]
```
You can put that just in one line, like
```
  skip: true  # [not linux]
```

marcelotrevisani commented on recipes/pynumero_build/meta.yaml:
```diff
@@ -0,0 +1,42 @@
+    
+about:
+  home: https://github.com/santiagoropb/pynumero_build.git
+  license: BSD
```
Please, add the ``license_file``

santiagoropb replied:
Sure I will add the file. I have a question about the output files in te recipe. My project builds two shared libraries and an executable
libpynumero_ASL.so
libpynumero_SPARSE.so
asl_test
CI is building correctly the libraries and the executable but conda is not finding them and its giving the warning 
WARNING: No files or script found for output pynumero_build
Is there a way to specify in the recipe that I want those three files in the output?

Thanks!

marcelotrevisani replied:
An easy way is just to copy your shared objects into ``$PREFIX/lib`` in your build script

marcelotrevisani replied:
Just the second one may be a problem

Try to add ``ampl-mp`` in run requirements.

But I think that's fine. I will mark @ocefpaf to also take a look on that

ocefpaf replied:
Let's call the experts @isuruf and @msarahan here :smile:

santiagoropb replied:
Thanks Marcelo and Filipe,

Initially I was planning on distributing the libraries by precompiling them so I am compiling with -static-libstdc++. I wonder if I can get rid of that flag now and let conda export the libc++ together with my libraries.

isuruf replied:
Both warnings can be ignored since we are using toolchain for now.

santiagoropb replied:
I just added the license file to the source. Seen that the build is passing I believe the pull request can be merged?

marcelotrevisani commented on recipes/pynumero_build/build.sh:
```diff
@@ -0,0 +1,20 @@
+mkdir build
+cd build
+echo "LOOKING FOR ampl-mp"
+find /opt/conda/pkgs -type d -name "ampl-mp*"
```
I think you have to modify this part. Just put ``pkg-config`` as a build dependency and use it directly instead of doing that

santiagoropb replied:
I am thinking on writing a cmake.find module in the future. The project uses that directory to link our compiled libraries to a library in ampl-mp. I am new to conda-forge so I am not sure if thats a robust way of doing it but for now seems to work for me.

marcelotrevisani commented on recipes/pynumero_build/build.sh:
```diff
@@ -0,0 +1,20 @@
+cd build
+echo "LOOKING FOR ampl-mp"
+find /opt/conda/pkgs -type d -name "ampl-mp*"
+mp_dir=$(find /opt/conda/pkgs -type d -name "ampl-mp*")
```
same here

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pynumero_build```) and found some lint.

Here's what I've got...


For **recipes/pynumero_build**:

 * The requirements section contained an unexpected subsection name. run_exports is not a valid subsection name.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pynumero_build```) and found it was in an excellent condition.

santiagoropb commented:
@marcelotrevisani do you think we can merge the pull request?

marcelotrevisani commented on recipes/pynumero_build/build.sh:
```diff
@@ -0,0 +1,34 @@
+#    find . -name "makefile" -exec sed -ie 's/CFLAGS = -DNo_dtoa -fPIC -O/CFLAGS = -fPIC -O/g' {} \;
+#fi
+#make
+#cd ../../../
```
Please remove the unnecessary commented code

santiagoropb replied:
Done

santiagoropb commented on recipes/pynumero_build/build.sh:
```diff
@@ -0,0 +1,34 @@
+echo "LOOKING FOR ampl-mp"
+if [[ "$OSTYPE" == "linux-gnu" ]]; then
+    find /opt/conda/pkgs -type d -name "ampl-mp*"
+    mp_dir=$(find /opt/conda/pkgs -type d -name "ampl-mp*")
```
I am thinking on writing a cmake.find module in the future. The project uses that directory to link our compiled libraries to a library in ampl-mp. I am new to conda-forge so I am not sure if thats a robust way of doing it but for now seems to work for me.

ocefpaf replied:
That is quite unorthodox and the hardcoded path is worrisome.
(I haven't really read all of your `build.sh` to understand why this is needed though.)

isuruf commented:
Build scripts in pynumero_libraries are a mess.

1. It links to a static library of libmp. (This needs to be fixed before we can merge this request)
2. sparse_utils has SHARED in it. This prevents the user from overriding it.
3. There's no install script. This makes the user install files manually.

Other things needed to do before merging this PR.

1. Build shared libraries. (Add `-DBUILD_SHARED_LIBS` to CMake line)
2. Remove run_exports line. I don't know why you would want that.

santiagoropb commented:
> 2\. t

Thanks @isuruf for your comments. My apologies for the mess. I am actually learning cmake. I tried to solve all the comments:

1. Pynumero_ASL is no longer linking to a static library in ampl-mp but to a shared library.
2. The end use of this libraries is just for linking (extension module in python) so I would not expect this to be a problem.
3. I attempted to pass install the libraries with CMAKE_INSTALL_PREFIX, however it is not placing the libraries in the right place which led me to add the copying statements manually

Other things:

1. Added the flag
2. Removed the run_exports keyword

isuruf commented:
For 3. you have a fix at 1bc2fb51e1aceea6e5a4a5ad0c5b9fe5f7929d22, but it's not used here

santiagoropb commented:
> For 3. you have a fix at 1bc2fb51e1aceea6e5a4a5ad0c5b9fe5f7929d22, but it's not used here

Thanks for the help! make install is now working.

santiagoropb commented:
@marcelotrevisani are we good for merging?

santiagoropb commented:
Thanks @marcelotrevisani! Can the recipe be now find in feedstock? I was trying to find it but I had no luck locating it. Also, once it is merged into feedstock what is the command to conda install it?
